tasks/main.yml

```diff
@@ -63,9 +63,55 @@
     name: dell-system-update
     state: latest
 
+- name: Init dell_dsu_update_all_firmware
+  set_fact:
+    dell_dsu_update_all_firmware: False
```

Comment:
I think this breaks workflow for users who have standardized on running stuff with CLI args `-e dell_dsu_update_all_firmware=True`.

I'm trying to wrap my brain around this. In general I think this functionality could be implemented *without* overriding the variable with set_fact, and primarily do what the user wants and what the user is used to, rather than introducing a lot of changes in the usage patterns of the role.

Reply:
can you just use inline magic when defaulting `dell_dsu_arguments: "--non-interactive --apply-upgrades"`?
you know like `dell_dsu_arguments: "--non-interactive --apply-upgrades {{ categories_list if categories_list is defined }}"`
with some more magic to make the list a comma seperated string with the right option in front of it? That way you can still run checks on the list if defined but keep the logic of the role mostly the same if it's not used

Reply:
It is not possible to use this dell_dsu_arguments: "--non-interactive --apply-upgrades {{ categories_list if categories_list is defined }}". The user will include names, such as BIOS,iDRAC...  Then, these names should be converted to symbols BI,DR... using the dictionary firmware_components.

Reply:
well you could also do a transformation in the default definition it will become fairly difficult to read inline magic then :) but you could also just have the users specify the proper component names for dsu without transforming them. I mean to me as a user there is not much of a difference of specifying iDRAC or DR directly.

Reply:
`dell_dsu_arguments: "--non-interactive --apply-upgrades{% if dsu_categories is defined %} --category={% set comma = joiner(',') %}{% for dsu_category in dsu_categories %}{{ comma() }}{{ firmware_components[dsu_category] }}{% endfor %}{% endif %}"`
something like this :)

Reply:
Thanks. I will check it by tomorrow and see.

Reply:
I have implemented your suggestions. Now, the code is shorter :).

Reply:
you can even put it into defaults/main.yml that would make it even easier and you'd need less checks. You only need to move the firmware_components definition in front of the dell_dsu_arguments definition

Reply:
Sorry, I did not get it. I need  categories_list to be defined before creating dell_dsu_arguments. In some cases, we need just to install the dsu without any updates.

Reply:
I think we're having a mismatch on what you want to achieve :) I thought you only want to add the possibility to upgrade certain types of firmware. Having this role not execute dsu upgrade, only installing it, was not on my scope

What I would do - create two input variables:
dsu_install_firmware (or something like that, default it to true)
dsu_categories, list of firmware components you want to upgrade, undefined by default. Can be defined by user as BIOS,iDRAC etc needs to be a list

Now you can use defaults/main.yml to do most of the work, something like what I already posted
```
dell_dsu_arguments: "--non-interactive --apply-upgrades{% if dsu_categories is defined %} --category={% set comma = joiner(',') %}{% for dsu_category in dsu_categories %}{{ comma() }}{{ firmware_components[dsu_category] }}{% endfor %}{% endif %}"
dsu_install_firmware: True
```

And in tasks/main.yml you only need to change the when clause of the update to 
`when: dsu_install_firmware | bool`

* this will keep the default behavior of update all firmware when dsu_categories is not supplied (ie no --category on dsu cli)
* it will change behavior when dsu_categories  is supplied to pass  --category=BI,... to dsu cli
* if you overwrite dsu_install_firmware: False it will not run dsu

I did not syntax check it, but at least I think it should work and minimize the changes you need to make :)

Reply:
Great. I will see how I can implement this changes.

Reply:
We are trying to update the code while keeping original code. As the code is already used by many users. 

For more context:

1) The code should install only dsu if no parameter has been defined. The ansible playbook that use this role, it sets dell_dsu_update_all_firmware to False. 

2) The code should update all the firmware if one of the following holds: i) dell_dsu_update_all_firmware=True; ii) upgrade_categories="all" 

3) The code should update only some firmware components if upgrade_categories is not set to all and has been defined with some components.  We have to assert to ensure that the passed parameters are valid. For example if we plan to update only BIOS and iDRAC, we just specify upgrade_categories="iDRAC,BIOS".

I think that the code that we have now is optimized for these use-cases and keep the old version working for others.

Reply:
I think this makes the role more complicated :) Anyways, I think the current code should work, but I think you need to check if you are making a breaking change, I think you can no longer overwrite dell_dsu_arguments which you could before your change (because you use set_fact to set dell_dsu_arguments)

Reply:
Before my change. dell_dsu_arguments has a fixed only one value, which is "--non-interactive --apply-upgrades". Now, it has more granularity. 

In the current code, we have dell_dsu_update_all_firmware = True. So, thanks to tasks defined from line 66 to 92, dell_dsu_arguments will take the value "--non-interactive --apply-upgrades" by default. Otherwise, if upgrade_categories is defined with more granularity, such as upgrade_categories="BIOS,iDRAC", then  dell_dsu_arguments will be fixed to "--non-interactive --apply-upgrades --category=BI,DR".

Reply:
> Before my change. dell_dsu_arguments has a fixed only one value, which is "--non-interactive --apply-upgrades". Now, it has more granularity.

no, I think you're missing some ansible basics. you can overwrite anything in defaults/main.yml without any code in tasks/main.yml, so you could overwrite it before with _anything_ you want including additional categories etc. For example I used it to add auto accept of gpg keys before the role implemented that. 

tldr: everything in defaults/main.yml is not a fixed value, it is a default that you can overwrite when calling the role or via host vars or ....

Reply:
Yes. I know that. The default value can be overwritten within the task or by passed parameters.